compiler/rustc_codegen_llvm/src/type_of.rs

```diff
@@ -90,14 +93,16 @@ fn uncached_llvm_type<'a, 'tcx>(
 fn struct_llfields<'a, 'tcx>(
     cx: &CodegenCx<'a, 'tcx>,
     layout: TyAndLayout<'tcx>,
-) -> (Vec<&'a Type>, bool) {
+) -> (Vec<&'a Type>, bool, Option<Box<[u32]>>) {
     debug!("struct_llfields: {:#?}", layout);
     let field_count = layout.fields.count();
 
     let mut packed = false;
     let mut offset = Size::ZERO;
     let mut prev_effective_align = layout.align.abi;
     let mut result: Vec<_> = Vec::with_capacity(1 + field_count * 2);
+    let mut projection = vec![0; field_count];
```

Comment:
Can you rename `projection` to `field_remapping`?

Also, if you want to, you can avoid allocating *until* you hit `padding_used = true` (and then you don't even need the separate `bool`).

All you have to do is to initialize it to `Vec::with_capacity(field_count)` + `.extend(0..i)` (when the `i`th field needs padding and it hasn't been initialized yet), after which you just push to it if it's `Some`.

Reply:
I did the rename.

Avoiding the allocation like this does not work though. We are iterating over the fields in memory order, not in source order, so `.extend(0..i)` does not generally give correct results.